proto/cacheclient.proto

```diff
@@ -402,16 +402,93 @@ message _ListRemoveRequest {
 
 message _ListRemoveResponse {}
 
+// List fetch + retain index behavior
+// This follows redis behavior for the most part, differences
+// 1. end_index is exclusive, unless None
+// 2. `start_index` == 0, then we start at the front of list
+// 2. `end_index` == None, start at the end of the index, inclusive
+/*
+  list_ = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+  # Using the slice notation:
+  # start:end
+
+
+  # start=0, end=None
+  list_[0:] == [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
+
+
+  # start=5, end=None
+  list_[5:] == [5, 6, 7, 8, 9]
+
+  # start=5, end=0
+  list_[5:0] == []
+
+  # start=-5, end=None
+  list_[-5:] == [5, 6, 7, 8, 9]
+
+  # start=-5, end=0
+  list_[-5:0] == []
+
+
+  # start=0, end=5
+  list_[0:5] == [0, 1, 2, 3, 4]
+
+  # start=0, end=5
+  list_[0:5] == [0, 1, 2, 3, 4]
+
+  # start=0, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+  # start=0, end=-5
+  list_[0:-5] == [0, 1, 2, 3, 4]
+
+
+  # start=2, end=6
+  list_[2:6] == [2, 3, 4, 5]
+
+  # start=2, end=-6
+  list_[2:-6] == [2, 3]
+
+  # start=-7, end=-2
+  list_[-7:-2] == [3, 4, 5, 6, 7]
+
+  // if using ltrim, this will remove all elements in the list
+  # start=-2, end=-6
+  list_[2:-6] == []
+
+  # empty list
+  list_[-1:1] == []
+ */
+message Unbounded { }
+
 message _ListFetchRequest {
   bytes list_name = 1;
   // Inclusive. 0 (start of list) by default.
   // A negative index counts from the end of the list
   sint32 start_index = 2;
-  // If unset or 0, this effectively means list.length()
+  // Exclusive if set.
+  // If unset, inclusive, this effectively means list.length()
   // If count is > the number of elements to return, return as much as you can
-  uint32 count = 3;
+  oneof end_index {
+    Unbounded unbounded = 3;
+    sint32 inclusive = 4;
+  }
```

Comment:
Looks good. We'll want the same pattern for `start_index`.

Reply:
Not necessary, the default value is 0 if `start_index` isn't included in the request, which is what we want. 

Start index can all just be 0, but for end index there's a difference between 0 and None, 0 being start of array and None being end of list.

Reply:
For future compatibility, if we support reverse order, then an omitted `start_index` will mean "end of the list". The client library can set `start_index` to -1 in that case (end of list inclusive), but it's more logic to put in the clients.